In [1]:
!pip install langchain
!pip install google-generativeai
!pip install InstructorEmbedding
!pip install langchain
!pip install transformers

!pip install -r requirements.txt

!pip install InstructorEmbedding
!pip install sentence-transformers


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 10.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.4/44.4 kB 5.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.4/49.4 kB 6.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 133.2/133.2 kB 3.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 267.9/267.9 kB 16.6 MB/s eta 0:00:00


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.7/7.7 MB 54.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.0/302.0 kB 37.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 108.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 97.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 295.0/295.0 kB 37.9 MB/s eta 0:00:00
ERROR: Could not open requirements file: [Errno 2] No such file or directory: 'requirements.txt'
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 86.0/86.0 kB 1.8 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 9.2 MB/s eta 0:00:00
  Created wheel for sentence-transformers: filename=sentence_transformers-2.2.2-py3-none-any.whl size=125923 sha256=a2ec633a2659b22a261a9b846c3a152706797dd7281cbfeb5e7a7e22e3c47d8b
  Stored in directory: /root/.cache/pip/wheels/62/f2/10/1e606fd5f02395388f74e7462910fe851042f97238cbbd902f
Successfully

In [2]:
import warnings as w
w.filterwarnings('ignore')

from langchain.llms import GooglePalm


from langchain.chains import RetrievalQA

from langchain.embeddings import GooglePalmEmbeddings
from langchain.embeddings import HuggingFaceInstructEmbeddings

from langchain.document_loaders.csv_loader import CSVLoader

from InstructorEmbedding import INSTRUCTOR

from langchain.vectorstores import FAISS


In [4]:
# Open the text file for reading
with open('api-key.txt', 'r') as file:
    # Read the contents of the file into a variable
    api_file = file.read()


In [5]:
api_key = api_file

In [6]:
llm = GooglePalm(google_api_key=api_key, temperature=0.1)

# Now let's load data from your file


In [7]:
import pandas as pd
df = pd.read_csv('/content/data.csv',encoding='cp1252')
df.head()

prompt  \
0  I have never done programming in my life. Can ...   
1                     Why should I trust Codebasics?   
2  Is there any prerequisite for taking this boot...   
3  What datasets are used in this bootcamp? Is it...   
4  I’m not sure if this bootcamp is good enough f...   

                                            response  
0  Yes, this is the perfect bootcamp for anyone w...  
1  Till now 9000 + learners have benefitted from ...  
2  Our bootcamp is specifically designed for begi...  
3  The datasets used in this bootcamp are crafted...  
4  We got you covered. Go ahead and watch our you...

In [8]:
loader = CSVLoader(file_path='/content/data.csv', source_column="prompt",encoding='cp1252')

# Store the loaded data in the 'data' variable
data = loader.load()


In [9]:
data[0]

Document(page_content='prompt: I have never done programming in my life. Can I take this bootcamp?\nresponse: Yes, this is the perfect bootcamp for anyone who has never done coding and wants to build a career in the IT/Data Analytics industry or just wants to perform better in your current job or business using data.', metadata={'source': 'I have never done programming in my life. Can I take this bootcamp?', 'row': 0})

# Hugging Face Embeddings

`We introduce Instructor👨‍🏫, an instruction-finetuned text embedding model that can generate text embeddings tailored to any task (e.g., classification, retrieval, clustering, text evaluation, etc.) and domains (e.g., science, finance, etc.) by simply providing the task instruction, without any finetuning.`
### Use cases
**`1. Calculate Sentence similarities`**

You can further use the model to compute similarities between two groups of sentences, with customized embeddings.

**`2. Information Retrieval`**

You can also use customized embeddings for information retrieval.

**`3. Clustering`**

Use customized embeddings for clustering texts in groups.



In [10]:
# Initialize instructor embeddings using the Hugging Face model
instructor_embeddings = HuggingFaceInstructEmbeddings(
    model_name="hkunlp/instructor-large",
     query_instruction="Represent the query for retrieval: ")


load INSTRUCTOR_Transformer
max_seq_length  512


In [11]:
e = instructor_embeddings.embed_query("What is your refund policy?")
len(e)

768

In [12]:
e[:10]

[-0.04209406301379204,
 0.008169054053723812,
 0.00031348480843007565,
 0.019235726445913315,
 0.03759321942925453,
 0.0623537115752697,
 0.01414873544126749,
 0.024791430681943893,
 -0.024264970794320107,
 0.03154788166284561]

As you can see above, embedding for a sentance "What is your refund policy" is a list of size 768. Looking at the numbers in this list, doesn't give any intuitive understanding of what it is but just assume that these numbers are capturing the meaning of "What is your refund policy". If you are curious to know about embeddings, go to youtube and search "codebasics word embeddings" and you will find bunch of videos with simple, intuitive explanations

# Vector store using FAISS

## `FAISS`
Faiss is a library for efficient similarity search and clustering of dense vectors. It contains algorithms that search in sets of vectors of any size, up to ones that possibly do not fit in RAM. It also contains supporting code for evaluation and parameter tuning. Faiss is written in C++ with complete wrappers for Python/numpy. Some of the most useful algorithms are implemented on the GPU.

In [13]:
!pip install faiss-gpu

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 85.5/85.5 MB 10.8 MB/s eta 0:00:00


In [14]:
# Create a FAISS instance for vector database from 'data'
vectordb = FAISS.from_documents(documents=data,
                                 embedding=instructor_embeddings)


In [15]:
# Create a retriever for querying the vector database
retriever = vectordb.as_retriever(score_threshold = 0.7)

In [16]:
rdocs = retriever.get_relevant_documents("how about job placement support?")
rdocs

[Document(page_content='prompt: Do you provide any job assistance?\nresponse: Yes, We help you with resume and interview preparation along with that we help you in building online credibility, and based on requirements we refer candidates to potential recruiters.', metadata={'source': 'Do you provide any job assistance?', 'row': 11}),
 Document(page_content='prompt: Will this course guarantee me a job?\nresponse: We created a much lighter version of this course on YouTube available for free (click this link) and many people gave us feedback that they were able to fetch jobs (see testimonials). Now this paid course is at least 5x better than the YouTube course which gives us ample confidence that you will be able to get a job. However, we want to be honest and do not want to make any impractical promises! Our guarantee is to prepare you for the job market by teaching the most relevant skills, knowledge & timeless principles good enough to fetch the job.', metadata={'source': 'Will this 

In [17]:

from langchain.chains import RetrievalQA

chain = RetrievalQA.from_chain_type(llm=llm,
                            chain_type="stuff",      # stuff chain type
                            retriever=retriever,   # database of embedding
                            input_key="query",
                            return_source_documents=True)

In [18]:
chain('how about job placement support?')

ServiceUnavailable: ignored

Many data have not persent in csv file but models given the answer on this capabilty but this is not good.....then use the Prompt Template and use it

# Create RetrievalQA chain along with `Prompt Template` 🚀

## `Prompt Template`
Prompt templates are pre-defined recipes for generating prompts for language models.

A template may include instructions, few-shot examples, and specific context and questions appropriate for a given task.

LangChain provides tooling to create and work with prompt templates.

LangChain strives to create model agnostic templates to make it easy to reuse existing templates across different language models.

Typically, language models expect the prompt to either be a string or else a list of chat messages.


In [ ]:
from langchain.prompts import PromptTemplate

prompt_template = """Given the following context and a question, generate an answer based on this context only.
In the answer try to provide as much text as possible from "response" section in the source document context without making much changes.
If the answer is not found in the context, kindly state "I don't know." Don't try to make up an answer.

CONTEXT: {context}

QUESTION: {question}"""


PROMPT = PromptTemplate(
    template=prompt_template, input_variables=["context", "question"]
)
chain_type_kwargs = {"prompt": PROMPT}


from langchain.chains import RetrievalQA

chain = RetrievalQA.from_chain_type(llm=llm,
                            chain_type="stuff",
                            retriever=retriever,
                            input_key="query",
                            return_source_documents=True,
                            chain_type_kwargs=chain_type_kwargs)
